File: throwcode_python/algorithm/sort/test_sort.py
#!/use/bin/env/python3
# -*- coding:utf-8 -*-
# @author: goinhn
# @date: 2021/3/2 11:13

import random
import algorithm.sort.bucket_sort as test_sort_single

"""
对数器测试排序算法
"""


def comparator(arr):
    list.sort(arr)


def generate_random_array(max_size, max_value):
    arr = [None] * random.randint(1, max_size)
    for i in range(len(arr)):
        arr[i] = random.randint(0, max_value) - random.randint(0, max_value)
    return arr


def copy_array(arr):
    if arr is None:
        return
    res = [None] * len(arr)
    for i in range(len(arr)):
        res[i] = arr[i]
    return res


def is_equal(arr1, arr2):
    if arr1 is None or arr2 is None:
        return False
    if len(arr1) != len(arr2):
        return False
    for i in range(len(arr1)):
        if arr1[i] != arr2[i]:
            return False
    return True


def print_array(arr):
    if arr is None:
        return
    print(arr)


def test_main(test_time, max_size, max_value):
    succeed = True
    for i in range(test_time):
        arr1 = generate_random_array(max_size, max_value)
        arr2 = copy_array(arr1)
        test_sort_single.bucket_sort(arr1)
        comparator(arr2)
        if not is_equal(arr1, arr2):
            succeed = False
            break
    print('Nice!' if succeed else 'Fucking!')


def main():
    test_main(10000, 100, 100)


if __name__ == '__main__':
    main()

File: throwcode_java/src/main/java/com/goinhn/throwcode/algorithm/sort/TestSort.java
package com.goinhn.throwcode.algorithm.sort;

import java.util.Arrays;

/**
 * 排序测试
 *
 * @author goinhn
 * @date 2021-03-01T17:10:38
 */
public class TestSort {

    /**
     * 内置api排序
     */
    public static void comparator(int[] arr) {
        Arrays.sort(arr);
    }

    /**
     * 生成随机的数组
     */
    public static int[] generateRandomArray(int maxSize, int maxValue) {
        int[] arr = new int[(int) ((maxSize + 1) * Math.random())];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = (int) ((maxValue + 1) * Math.random()) - (int) (maxValue * Math.random());
        }
        return arr;
    }

    /**
     * 拷贝数组
     */
    public static int[] copyArray(int[] arr) {
        if (arr == null) {
            return null;
        }
        int[] res = new int[arr.length];
        System.arraycopy(arr, 0, res, 0, arr.length);
        return res;
    }

    /**
     * 判断数组是否相等
     */
    public static boolean isEqual(int[] arr1, int[] arr2) {
        if (arr1 == null || arr2 == null) {
            return false;
        }
        if (arr1.length != arr2.length) {
            return false;
        }
        for (int i = 0; i < arr1.length; i++) {
            if (arr1[i] != arr2[i]) {
                return false;
            }
        }
        return true;
    }

    /**
     * 打印数组
     */
    public static void printArray(int[] arr) {
        if (arr == null) {
            return;
        }
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void testMain(int testTime, int maxSize, int maxValue) {
        boolean succeed = true;
        for (int i = 0; i < testTime; i++) {
            int[] arr1 = generateRandomArray(maxSize, maxValue);
            int[] arr2 = copyArray(arr1);
            printArray(arr1);
            new QuickSort.ClassicQuickSort().quickSort(arr1);
            comparator(arr2);
            if (!isEqual(arr1, arr2)) {
                succeed = false;
                break;
            }
        }
        System.out.println(succeed ? "Nice!" : "Fucking fucked!");
    }


    public static void main(String[] args) {
        testMain(500000, 10, 10);
    }

}
